### PR TITLE
ENH, CI: Add Emscripten to CI

### PR DESCRIPTION
**Pull request by hoodmane**
This is using a branch of Pyodide that finishes support for out of tree builds. This is blocked on Pyodide:
- [x] Merge changes from my branch into main Pyodide
- [x] Release Pyodide v0.21.0a3 (hopefully) with out of tree build support

Other than that, I think it is nearly ready.

### CODE REVIEW

File: numpy/core/src/multiarray/datetime.c
Comment (mattip) [inline numpy/core/src/multiarray/datetime.c:3908]: Is this change intentional?

Comment (seberg) [reply numpy/core/src/multiarray/datetime.c:3908]: The change is correct, but was split out into its seperate PR already.

**Comment by mattip** [comment]
The lint failures due to too-long-lines are OK, but could you clean up the too-many-blank-lines ones? 

Any idea where the CI run of this is, I don't seem to see it in the summary.

File: emscripten/emscripten_runner.js
Comment (rth) [inline emscripten/emscripten_runner.js:60]: Cool, so did you manage to get live output from pytest with this?

File: emscripten/emscripten_runner.js
Comment (rth) [inline emscripten/emscripten_runner.js:1]: I think ideally we should ship this setup in https://github.com/pyodide/pyodide/tree/main/pyodide-test-runner so one could just pip install it and not have this file in each tested repo.

File: .github/workflows/emscripten.yml
Comment (rth) [inline .github/workflows/emscripten.yml:17]: Maybe `ubuntu-22.04` as I'm not sure this is forward compatible?

Comment (hoodmane) [reply .github/workflows/emscripten.yml:17]: Most of the other ci jobs run in ubunti-latest though there are a few that are version pinned. We can leave it up to the numpy maintainers. @mattip @seberg opinions?

Comment (seberg) [reply .github/workflows/emscripten.yml:17]: I don't care.  We will probably ping you anyway if the job starts failing ;).

File: emscripten/emscripten_runner.js
Comment (hoodmane) [reply emscripten/emscripten_runner.js:60]: Yup, it almost works like a normal tty: it can update the same line, use color, etc. There is some trouble with flushing, I think the problem might be here:
https://github.com/python/cpython/blob/3.10/Modules/_io/textio.c#L1536-L1537
`sys.stdout.flush()` apparently doesn't actually flush file descriptor 1.

I think that a solution would be that shouldn't buffer writes to stdout.

It would be good to add something like this to Pyodide. Some issues:
1. By default, `sys.stdout.isatty()` is `True` but writing out ANSI color codes and terminal commands to `console.log` just prints nonsense. It should be false in these cases.
2. If you pass a custom `stdout` to `loadPyodide`, then `sys.stdout.isatty()` returns `False` and there is no way to ask it to return `True`.
3. Node applications can work correctly as ttys as (mostly) demonstrated here but it would be nice to offer that capability in some simple way for other node cli applications. Enabling this in tests breaks our node test runner though.

Most of these are also Emscripten issues. See related:
https://github.com/emscripten-core/emscripten/pull/17320

**Comment by hoodmane** [comment]
I updated this to use Pyodide v0.21.0. Would appreciate review.

File: .github/workflows/emscripten.yml
Comment (rgommers) [inline .github/workflows/emscripten.yml:6]: I guess we can leave this in as a convenient way to trigger a build on a fork perhaps, but not sure that that is intended?

File: .github/workflows/emscripten.yml
Comment (rgommers) [inline .github/workflows/emscripten.yml:16]: This needs a `if: github.repository == 'numpy/numpy'` condition just like all the other actions have - preventing it from running on forks by default. Although this action does not trigger on a push to `main`, so perhaps it can be left out here.

File: .github/workflows/emscripten.yml
Comment (rgommers) [reply .github/workflows/emscripten.yml:17]: Agreed, it's fine either way.

File: emscripten/emscripten_runner.js
Comment (rgommers) [inline emscripten/emscripten_runner.js:96]: Can you remove this line?

Comment (rgommers) [reply emscripten/emscripten_runner.js:96]: Pinning most of these deps seems fine to me.

File: numpy/core/tests/test_umath.py
Comment (rgommers) [inline numpy/core/tests/test_umath.py:3474]: This is the first non-specific reason. Do you expect it will remain non-working? If not, maybe open an upstream issue and link it here?

**Review by rgommers — COMMENTED** [review]
Thanks @hoodmane. This is a large diff, but overall it looks very reasonable to me - pretty much everything seems to work, except for:
- floating-point exceptions
- starting a subprocess
- threading
- invoking a compiler
- memmap

Those all make perfect sense. I assume that that's going to stay unchanged for quite a while?

**Comment by rgommers** [comment]
This does need a small rebase or merging in `main` to resolve the merge conflict.

**Comment by rgommers** [comment]
I'm looking through the build log, and seeing that `scipy` and `pythran` are pulled in, plus `setuptools-rust` is built:
```
Successfully built setuptools-rust
Installing collected packages: ply, typing_extensions, setuptools, semantic-version, pycparser, numpy, gast, setuptools-rust, scipy, cffi, beniget, pythran
Successfully installed beniget-0.4.1 cffi-1.15.0 gast-0.5.3 numpy-1.23.1 ply-3.11 pycparser-2.21 pythran-0.11.0 scipy-1.8.1 semantic-version-2.10.0 setuptools-64.0.3 setuptools-rust-1.3.1.dev24+g5e8c380 typing_extensions-4.3.0
```
Doesn't seem like those should be needed?

**Comment by rgommers** [comment]
No BLAS/LAPACK found, only:
```
INFO: numpy_linalg_lapack_lite:
INFO:   FOUND:
INFO:     language = c
```
Is that expected to remain the case? Not sure what the status is of porting a BLAS/LAPACK implementation to Pyodide.

**Comment by rgommers** [comment]
All SIMD optimizations get disabled:
```
WARN: CCompilerOpt.__init__[1012] : unable to detect CPU architecture which lead to disable the optimization. check dist_info:<<
('emscripten_3_1_14_wasm32', 'cc', '-Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -O2 -g0 -fPIC -g2')
>>
WARN: CCompilerOpt.__init__[1028] : unable to detect compiler type which leads to treating it as GCC. this is a normal behavior if you're using gcc-like compiler such as MinGW or IBM/XLC.check dist_info:<<
('emscripten_3_1_14_wasm32', 'cc', '-Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -O2 -g0 -fPIC -g2')
```
I assume that that's fine for now - SSE instruction support is very spotty still (and anything else non-existent it looks like): https://emscripten.org/docs/porting/simd.html#compiling-simd-code-targeting-x86-sse-instruction-set

**Comment by rgommers** [comment]
The build phase looks good. For the test phase, there's >500 lines of warnings under `warnings summary`. Would it be possible to get `emscripten_runner.js` or `pytest` to suppress that output?

**Comment by rgommers** [comment]
I reviewed everything except `emscripten_runner.js`; I think we can rely on @hoodmane and @rth's expertise for that file. Overall this looks close to merge-ready. I don't think we had a discussion about whether we want this CI job (or did we?), but I'm +1. It's still a little experimental right now, but I'd say we can merge and see how it goes. In case there are a lot of issues (not too likely it seems, given the limited number of reasons for test skips) we can always reconsider.

File: .github/workflows/emscripten.yml
Comment (hoodmane) [reply .github/workflows/emscripten.yml:6]: This was intended as a temporary thing, will revert before merging.

File: numpy/core/tests/test_umath.py
Comment (hoodmane) [reply numpy/core/tests/test_umath.py:3474]: I think there are two problems that are labeled as "doesn't work":
1. `arctan` and `arctanh` fail the branch cut tests
2. `long double` precision is sometimes very bad (worse than `double` precision)

I will go back through at some point and double check so we can give more specific reasons. I'm not sure what the likelihood of fixing this is, but it isn't a very high priority for me.

**Comment by hoodmane** [comment]
> I'm looking through the build log, and seeing that `scipy` and `pythran` are pulled in, plus `setuptools-rust` is built:
> ...
> Doesn't seem like those should be needed?

Currently the Pyodide build system has a bunch of hacks to get specific packages to work. Over time we should work on reducing them, but I am pretty willing to introduce hacks if they get things to work. I made a PR to remove the `setuptools-rust` dependency:
https://github.com/pyodide/pyodide/pull/3045

**Comment by hoodmane** [comment]
> No BLAS/LAPACK found
> ...
> Is that expected to remain the case? Not sure what the status is of porting a BLAS/LAPACK implementation to Pyodide.

We have a BLAS/LAPACK implementation based on CLAPACK (which is quite old: v3.2.1) + f2c applied to a few newer functions that scipy needs. We don't use it for numpy currently. I am not sure why, the decision not to use it in numpy predates me. Maybe @rth can explain.

In any case, it would add a bit more complexity to use it. Probably something to look into in the future. Again, it would be great to get a working fortran compiler and switch to LAPACK 3.10.

**Comment by hoodmane** [comment]
> The build phase looks good. For the test phase, there's >500 lines of warnings under warnings summary. Would it be possible to get emscripten_runner.js or pytest to suppress that output?

Mostly they are `PendingDeprecationWarning`. Do you have some standard way of suppressing them in the other tests?

**Comment by hoodmane** [comment]
> I reviewed everything except `emscripten_runner.js`

If we hold off on this a bit longer until the work in https://github.com/pyodide/pyodide/pull/2976 is completed, we can remove this script and use the following:
```sh
python -m venv .venv-host
source .venv-host/bin/activate
pip install pyodide-build
pyodide venv .venv-pyodide
source .venv-pyodide/bin/activate
pip install -r test_requirements.txt
# Inside the .venv-pyodide, the following will run tests in Pyodide =)
python -b ./runtests.py -n -v --mode=full --durations 10 --coverage
```
I am planning to release Pyodide v0.22.0a1 as soon as we merge the command line runner stuff.

**Comment by rth** [comment]
> Is that expected to remain the case? Not sure what the status is of porting a BLAS/LAPACK implementation to Pyodide.

There is a plan to start using BLIS https://github.com/pyodide/pyodide/issues/227 as it's probably easier to build for WASM than OpenBLAS. Not much progress for now, though I think it essentially works https://github.com/flame/blis/issues/491 without SIMD support.  Though of course, we are receptive if you have other suggestions about the BLAS we should use.

File: numpy/core/tests/test_umath.py
Comment (rgommers) [reply numpy/core/tests/test_umath.py:3474]: Makes sense, not super important indeed.

Comment (rgommers) [reply numpy/core/tests/test_umath.py:3474]: Do you know what the size of `long double` in WASM is? I did a quick search, and found https://github.com/emscripten-core/emscripten/issues/5996#issuecomment-355621435 - but that's 4 years old by now.

**Comment by rgommers** [comment]
Re BLAS/LAPACK, thanks, that all makes sense. Performance isn't critical yet I guess - first make it all work. BLIS does seem like a good choice for BLAS, and it'll work fine with NumPy and SciPy. libFlame is still a work in progress though I believe, and not moving fast. So it's BLIS + Netlib LAPACK vs. OpenBLAS. Still makes sense to go for the former, OpenBLAS is pretty hairy internally.

> If we hold off on this a bit longer until the work in [pyodide/pyodide#2976](https://github.com/pyodide/pyodide/pull/2976) is completed, we can remove this script and use the following:

Great! I'd also be fine with merging this PR as is and removing the `.js` file once it's no longer needed - whatever you prefer.

> Mostly they are `PendingDeprecationWarning`. Do you have some standard way of suppressing them in the other tests?

The warning filters in `pytest.ini` are supposed to take care of this. Also, `pytest` has command-line flags for this, see https://stackoverflow.com/questions/40710094/how-to-suppress-py-test-internal-deprecation-warnings

**Comment by hoodmane** [comment]
Huh, it works fine locally...

File: numpy/core/tests/test_umath.py
Comment (hoodmane) [reply numpy/core/tests/test_umath.py:3709]: Yeah I am pretty darn sure that it's 128 bits.

**Comment by hoodmane** [comment]
There's another function signature mismatch to fix in `test_generator_mt19937.py::TestSetState::test_negative_binomial`.

**Comment by hoodmane** [comment]
@rgommers any idea about the `test_negative_binomial` function pointer cast problem? It happens inside the `disc` function which seems to do a bunch of branching...
https://github.com/numpy/numpy/blob/main/numpy/random/_common.pyx#L845

**Comment by hoodmane** [comment]
Okay strangely if I run the test in browser rather than in node, it passes. Probably I will just xfail it because that's too weird.

File: .github/workflows/emscripten.yml
Comment (rth) [inline .github/workflows/emscripten.yml:65]: IKs there a typo in .patch?

File: tools/pyodide_python_patch.patc
Comment (rth) [inline tools/pyodide_python_patch.patc:2]: Maybe add a comment that this would be fixed in the next Pyodide release?

**Review by rth — COMMENTED** [review]
Very nice!

File: .github/workflows/emscripten.yml
Comment (hoodmane) [reply .github/workflows/emscripten.yml:65]: No, `*.patch` and `*.diff` are in the `.gitignore` file so I changed the name to not match that pattern. This has the disadvantage of looking like a typo though.

**Comment by hoodmane** [comment]
So the reproduction is:
```
from numpy.random import Generator, MT19937
seed = 1234567890
rg = Generator(MT19937(seed))
rg.negative_binomial(0.5, 0.5)
```
this fails in all versions of node but works fine in chrome and firefox...

**Comment by hoodmane** [comment]
```js
 Stack (most recent call first):
  <no Python frame>
ExitStatus {
  name: 'ExitStatus',
  message: 'Program terminated with exit(0)',
  status: 0
}
Error: Process completed with exit code 1.
```
So close...

**Comment by hoodmane** [comment]
Okay I think this is ready to merge.

File: .github/workflows/emscripten.yml
Comment (lesteve) [inline .github/workflows/emscripten.yml:3]: I guess this line should **not** be commented out

**Review by lesteve — COMMENTED** [review]
There are some linting errors (lines too long) that should be easy to fix, see [build log](https://github.com/numpy/numpy/actions/runs/3439998344/jobs/5737943258).

From the sideline (I am not a numpy maintainer), this is quite nice to see this moving forward again!

File: numpy/testing/_private/utils.py
Comment (rth) [inline numpy/testing/_private/utils.py:51]:
```suggestion
IS_WASM = platform.machine() in ["wasm32", "wasm64"]
```

File: .github/workflows/emscripten.yml
Comment (rth) [inline .github/workflows/emscripten.yml:20]: Maybe add a comment that these variables can only be changed together and link to 
https://github.com/pyodide/pyodide/blob/main/Makefile.envs#L2 where they are defined for a given Pyodide version.

**Review by rth — COMMENTED** [review]
Very nice, Hood! Great that you managed to make it work in the venv without the extra JS script.

The only thing I'm wondering about is whether it would be better to differentiate
 -  skiped test that cannot be fixed in WASM, e.g. subprocess will never work
 - xfail for things that are potentially fixable. For instance, the fp exception might be fixable, https://github.com/pyodide/pyodide/issues/156#issuecomment-804743342 is probably outdated now, but it's still something about FP exception handling in musl unless you see a fundamental reason why it wouldn't work.

The only unfortunate thing is `@pytest.mark.xfailif` doesn't exist.


As otherwise now it would show,
```
= 19491 passed, 2186 skipped, 1305 deselected, 37 xfailed, 2 xpassed in 370.63s (0:06:10) =
```
and in those 2k skipped tests it's a bit tedious to review what's skipped due to WASM and what to something else.

File: numpy/lib/tests/test_io.py
Comment (rth) [inline numpy/lib/tests/test_io.py:2543]: Similarly, maybe xfail this. Someone could try to report it to emscripten and see if it's fixable.

File: numpy/lib/tests/test_format.py
Comment (rth) [inline numpy/lib/tests/test_format.py:530]: Xfail

File: numpy/core/tests/test_umath.py
Comment (rth) [inline numpy/core/tests/test_umath.py:3808]: xfail also all of the above

File: numpy/core/tests/test_regression.py
Comment (rth) [inline numpy/core/tests/test_regression.py:329]: xfail

File: .github/workflows/emscripten.yml
Comment (hoodmane) [reply .github/workflows/emscripten.yml:3]: It's a confusing comment, but it is the same as the comments in all of the other workflows. It's referring to line 15.

File: .github/workflows/emscripten.yml
Comment (hoodmane) [reply .github/workflows/emscripten.yml:20]: Right. Except for `NODE_VERSION` which is more flexible. As a later refinement, it would be good to fill `PYTHON_VERSION` and `EMSCRIPTEN_VERSION` from the repodata.json file. Maybe we could make a github action that does the setup for a specific Pyodide version.

File: numpy/lib/tests/test_io.py
Comment (hoodmane) [reply numpy/lib/tests/test_io.py:2543]: We should double check but I think they know about the problem with memmap. I believe it is fixable but nobody has gotten to it.

File: numpy/lib/tests/test_format.py
Comment (hoodmane) [reply numpy/lib/tests/test_format.py:530]: I've spent so much time fixing dup bugs in various Emscripten file systems. It's disappointing to keep seeing more of them.

File: .github/workflows/emscripten.yml
Comment (lesteve) [reply .github/workflows/emscripten.yml:3]: Ah OK makes sense ...

**Comment by hoodmane** [comment]
> xfail for things that are potentially fixable. For instance, the fp exception might be fixable, https://github.com/pyodide/pyodide/issues/156#issuecomment-804743342 is probably outdated now, but it's still something about FP exception handling in musl unless you see a fundamental reason why it wouldn't work.

I think it's not a problem with musl floating point exceptions, but rather with wasm floating point exceptions. And the fact that they don't exist. Per the spec:

> All operators use “non-stop” mode, and floating-point exceptions are not otherwise observable. In particular, neither alternate floating-point exception handling attributes nor operators on status flags are supported. There is no observable difference between quiet and signalling NaNs.
https://webassembly.github.io/spec/core/exec/numerics.html#floating-point-operations

My belief is that many of the floating point tests are unfixable without either software emulating floating point exceptions (this would be a large amount of work and the result would be very slow) or until the wasm spec and vms are updated to relax these limitations. The spec does say:

> Note: Some of these limitations may be lifted in future versions of WebAssembly.

But note that there is no existing proposal to add fp exception support, so IMO such support is at least 3 years away and probably  further.

**Comment by hoodmane** [comment]
Thanks for the review @rth!

**Comment by hoodmane** [comment]
> The only unfortunate thing is `@pytest.mark.xfailif` doesn't exist.

I just looked this up and apparently `@pytest.mark.xfail` can be used this way:
https://docs.pytest.org/en/6.2.x/skipping.html#condition-parameter
But the reason I decided to skip all these tests was indeed because I didn't realize that `@pytest.mark.xfail` has a condition parameter.

**Review by rgommers — APPROVED** [review]
This looks very clean now, it's all green, and it's a pretty fast CI job. So let's get this in. I pushed one change to ensure we use the same permission guard and concurrency settings as in other CI jobs.

Thanks a lot @hoodmane & all reviewers!

**Comment by rgommers** [comment]
This job is now failing for the last 1-2 days because of:
```
Run actions/setup-python@v4
Version 3.10.2 was not found in the local cache
Error: Version 3.10.2 with arch x64 not found
The list of all available versions can be found here: https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json
```

Checking the link, `3.10.2` for `x86` is listed, so not sure what is up there. @hoodmane does this ring a bell? Is it safe to switch to a newer `3.10.x` version?

**Comment by hoodmane** [comment]
I think @ryanking13 fixed a similar issue on our repo. It should be fine to switch to a newer Python 3.10.x. After the switch to Python 3.11 we should maybe start doing patch updates...

**Comment by ryanking13** [comment]
> This job is now failing for the last 1-2 days because of:
> 
> ```
> Run actions/setup-python@v4
> Version 3.10.2 was not found in the local cache
> Error: Version 3.10.2 with arch x64 not found
> The list of all available versions can be found here: https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json
> ```
> 
> Checking the link, `3.10.2` for `x86` is listed, so not sure what is up there. @hoodmane does this ring a bell? Is it safe to switch to a newer `3.10.x` version?

ubuntu22.04 (ubuntu-latest) doesn't have python 3.10.2 available with setup-python actions. See list of available python versions here: https://github.com/actions/python-versions/blob/main/versions-manifest.json

I think you can either use ubuntu20.04 or use newer python 3.10.x.

**Pull request merged**
